### ISSUE TITLE
[QUESTION] What's the correct way to train XGBoost on a multi-source dataset?

### ISSUE THREAD

**Issue by giacomoguiduzzi** [unit8co/darts#2570]
Greetings,

I am trying to fit an `XGBoost` model using a dataset that is made of samplings from multiple data sources. Each data source is associated to an ID. The best way I found to represent this through `TimeSeries` instances is to stack the measurements on the sample axis using a common time axis. Giving the XGBoost model the resulting `TimeSeries` as-is works, but I don't want a probabilistic forecasting, I want a single-step one. In my data processing pipeline I created a sliding window dataset where each window is associated to a label at `t` timesteps in the future, so I tried creating a different `TimeSeries` object for each window; I get an exception about the size of the `TimeSeries` objects, and I think it is because these objects do not have the label in them (if I'm using a look-back window size of 96 for example, these TimeSeries objects are going to be of size 96, thus it raises an exception about arrays being 0-sized). Looking at the code for the `XGBoost` model, it looks like it works in a sort of auto-regressive mode: if I got it correctly, it is using the previous 96 values of the given `TimeSeries` to forecast what is coming immediately after that. Since I have a forecasting horizon n=1, does the model need `TimeSeries` objects of size 97? If this is correct, would the procedure of giving the model a sequence of `TimeSeries` objects work for training as it would for other models using the sliding window dataset? I am not sure the outcome is the same as passing a single dataset object.
I tried summing up the problem to avoid a wall of text, so let me know if you need any further information.
Thank you so much for your support. I'm looking forward to your kind response.

Best Regards,
Giacomo Guiduzzi

**Comment by madtoinou** [comment]
Hi @giacomoguiduzzi,

First of all, sorry for the delay, I remember reading your issue and then got submerged by other stuff and forgot about it.


> I am trying to fit an `XGBoost` model using a dataset that is made of samplings from multiple data sources. Each data source is associated to an ID. The best way I found to represent this through `TimeSeries` instances is to stack the measurements on the sample axis using a common time axis.

It sounds great, especially if those various sources are tracking the same process. Any reason you did not go for the multi-series approach, encoding the ID as a static covariates?

> Giving the XGBoost model the resulting `TimeSeries` as-is works, but I don't want a probabilistic forecasting, I want a single-step one.

Taking the multivariate approach is not incompatible with the deterministic forecast. You can just set `output_chunk_length=1` and call `predict()` with `n=1`.

> In my data processing pipeline I created a sliding window dataset where each window is associated to a label at `t` timesteps in the future, so I tried creating a different `TimeSeries` object for each window; I get an exception about the size of the `TimeSeries` objects, and I think it is because these objects do not have the label in them (if I'm using a look-back window size of 96 for example, these TimeSeries objects are going to be of size 96, thus it raises an exception about arrays being 0-sized).

This is a nice engineered feature, but you are indeed dependent on the length of the series.

> Looking at the code for the `XGBoost` model, it looks like it works in a sort of auto-regressive mode: if I got it correctly, it is using the previous 96 values of the given `TimeSeries` to forecast what is coming immediately after that. Since I have a forecasting horizon n=1, does the model need `TimeSeries` objects of size 97?

If the model is created with `lags=96`, it will use the 96 steps in the past to predict 1 step. In order to fit the model, the series probably need to be at least 98 steps-long in order to be able to generate 3 training samples. The fact that your target is the output of the sliding window mentioned above (if I understood correctly) does not impact the length requirements as the lags of the model and size of the window seems to be identical.

>  If this is correct, would the procedure of giving the model a sequence of `TimeSeries` objects work for training as it would for other models using the sliding window dataset? I am not sure the outcome is the same as passing a single dataset object. I tried summing up the problem to avoid a wall of text, so let me know if you need any further information. Thank you so much for your support. I'm looking forward to your kind response.

The only difference by using a sequence of `TimeSeries` instead of a multivariate one, is that the model will generalize to all these series (you should be able to increase the size of the model since more samples will be generated) instead of leveraging all the features across components (sources) to predict all the sources. It ultimately depends if the task of the model is to smooth the values recorded across the sources or predict each of the process independently.

Sorry again for the delay, I hope that my answer is helpful.

**Comment by giacomoguiduzzi** [comment]
Hi @madtoinou ,

Thank you very much for your insight. For now I don't need to continue with my attempt at XGBoost, but you gave me very useful information about handling data for other models. What might not have been clear is that the dataset is multivariate, but also has IDs; this means that I have an array of 2D data. I might use the ID as a covariate as you suggested, but I don't want the model to use the ID as information. I'd like to exploit a similar approach I used for other models through a custom dataloader class which serves previously processed windows.

Have a great day ahead!
Giacomo
